packages/react-dom/src/__tests__/ReactDOMFizzShellHydration-test.js

```diff
@@ -283,4 +283,31 @@ describe('ReactDOMFizzShellHydration', () => {
     ]);
     expect(container.textContent).toBe('New screen');
   });
+
+  test('TODO: A large component stack causes SSR to stack overflow', async () => {
+    spyOnDevAndProd(console, 'error');
+
+    function createNestedComponent(depth: number) {
+      if (depth <= 0) {
+        return function Leaf() {
+          return <AsyncText text="Shell" />;
+        };
+      }
+      const NextComponent = createNestedComponent(depth - 1);
+      function Component() {
+        return <NextComponent />;
+      }
+      return Component;
+    }
+    const NestedComponent = createNestedComponent(1500);
```

Comment:
I think you can just make this a single component. Your current implementation returns a different instance on every re-render even for the same depth which is maybe not ideal.

```suggestion
    function NestedComponent({depth}: {depth: number}) {
      if (depth <= 0) {
        return null;
      }
      return <NestedComponent depth={props.depth - 1} />;
    }
```

Reply:
Great catch!